emscripten.py

```diff
@@ -343,39 +343,38 @@ def create_fp_accessors(metadata):
 
 
 def create_named_globals(metadata):
-  if not shared.Settings.RELOCATABLE:
-    named_globals = []
-    for k, v in metadata['namedGlobals'].items():
+  named_globals = []
+  for k, v in metadata['namedGlobals'].items():
+    v = int(v)
+    if shared.Settings.RELOCATABLE:
+      v += shared.Settings.GLOBAL_BASE
+    elif k == '__data_end':
       # We keep __data_end alive internally so that wasm-emscripten-finalize knows where the
       # static data region ends.  Don't export this to JS like other user-exported global
       # address.
-      if k not in ['__data_end']:
-        named_globals.append("Module['_%s'] = %s;" % (k, v))
-    return '\n'.join(named_globals)
-
-  elements = ',\n  '.join('"' + k + '": ' + str(v) for k, v in metadata['namedGlobals'].items())
-  named_globals = '''
-var NAMED_GLOBALS = {
-  %s
-};
-for (var named in NAMED_GLOBALS) {
-  Module['_' + named] = %s + NAMED_GLOBALS[named];
-}
-Module['NAMED_GLOBALS'] = NAMED_GLOBALS;
-''' % (elements, shared.Settings.GLOBAL_BASE)
-
-  # wasm side modules are pure wasm, and cannot create their g$..() methods, so we help them out
-  # TODO: this works if we are the main module, but if the supplying module is later, it won't, so
-  #       we'll need another solution for that. one option is to scan the module imports, if/when
-  #       wasm supports that, then the loader can do this.
-  named_globals += '''
-for (var named in NAMED_GLOBALS) {
-  (function(named) {
-    var addr = Module['_' + named];
-    Module['g$_' + named] = function() { return addr };
-  })(named);
+      continue
+    mangled = asmjs_mangle(k)
+    if shared.Settings.MINIMAL_RUNTIME:
+      named_globals.append("var %s = %s;" % (mangled, v))
+    else:
+      named_globals.append("var %s = Module['%s'] = %s;" % (mangled, mangled, v))
```

Comment:
do we need the export on `Module` here? I think only for relocatable since the dynamic linker looks there?

(otherwise, I'd guess we should export on Module only stuff the user asked to export)

Reply:
Do you count EMSCRIPTEN_KEEPALIVE symbols when you say "only stuff the user asked to export" ?  Is EMSCRIPTEN_KEEPALIVE supported to export on the module too?

Reply:
It does currently. So if this code path is for that, then sgtm.

Reply:
I think we could find a way to perhaps limit the nameGlobals that we choose to assign to the Module, but its tricky because we don't know the list of EMSCRIPTEN_KEEPALIVE functions so we can't just ignore things not in USER_EXPORTED_FUNCTIONS.

I guess we could do EXPORTED_FUNCTIONS - USER_EXPORTED_FUNCTIONS to find all the internal symbols that were exported by system code and exlude those?  But we don't currently do that for function symbols either.

I'd rather do that in a followup for both data and function symbols I think.   Here in this change I'm mostly focused here on removing the `NAMED_GLOBALS` variable.